File: 研究/cli-main/cli-main/internal/auth/device_flow.go
// Copyright (c) 2026 Lark Technologies Pte. Ltd.
// SPDX-License-Identifier: MIT

package auth

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/larksuite/cli/internal/core"
)

// DeviceAuthResponse is the response from the device authorization endpoint.
type DeviceAuthResponse struct {
	DeviceCode              string `json:"device_code"`
	UserCode                string `json:"user_code"`
	VerificationUri         string `json:"verification_uri"`
	VerificationUriComplete string `json:"verification_uri_complete"`
	ExpiresIn               int    `json:"expires_in"`
	Interval                int    `json:"interval"`
}

// DeviceFlowTokenData contains the token data from a successful device flow.
type DeviceFlowTokenData struct {
	AccessToken      string
	RefreshToken     string
	ExpiresIn        int
	RefreshExpiresIn int
	Scope            string
}

// DeviceFlowResult is the result of polling the token endpoint.
type DeviceFlowResult struct {
	OK      bool
	Token   *DeviceFlowTokenData
	Error   string
	Message string
}

// OAuthEndpoints contains the OAuth endpoint URLs.
type OAuthEndpoints struct {
	DeviceAuthorization string
	Token               string
}

// ResolveOAuthEndpoints resolves OAuth endpoint URLs based on brand.
func ResolveOAuthEndpoints(brand core.LarkBrand) OAuthEndpoints {
	ep := core.ResolveEndpoints(brand)
	return OAuthEndpoints{
		DeviceAuthorization: ep.Accounts + "/oauth/v1/device_authorization",
		Token:               ep.Open + "/open-apis/authen/v2/oauth/token",
	}
}

// RequestDeviceAuthorization requests a device authorization code.
func RequestDeviceAuthorization(httpClient *http.Client, appId, appSecret string, brand core.LarkBrand, scope string, errOut io.Writer) (*DeviceAuthResponse, error) {
	if errOut == nil {
		errOut = io.Discard
	}

	endpoints := ResolveOAuthEndpoints(brand)

	if !strings.Contains(scope, "offline_access") {
		if scope != "" {
			scope = scope + " offline_access"
		} else {
			scope = "offline_access"
		}
	}

	basicAuth := base64.StdEncoding.EncodeToString([]byte(appId + ":" + appSecret))

	form := url.Values{}
	form.Set("client_id", appId)
	form.Set("scope", scope)

	req, err := http.NewRequest("POST", endpoints.DeviceAuthorization, strings.NewReader(form.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Authorization", "Basic "+basicAuth)

	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("Device authorization failed: read body: %v", err)
	}

	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, fmt.Errorf("Device authorization failed: HTTP %d – response not JSON", resp.StatusCode)
	}

	_, hasError := data["error"]
	if resp.StatusCode >= 400 || hasError {
		msg := getStr(data, "error_description")
		if msg == "" {
			msg = getStr(data, "error")
		}
		if msg == "" {
			msg = "Unknown error"
		}
		return nil, fmt.Errorf("Device authorization failed: %s", msg)
	}

	expiresIn := getInt(data, "expires_in", 240)
	interval := getInt(data, "interval", 5)

	verificationUri := getStr(data, "verification_uri")
	verificationUriComplete := getStr(data, "verification_uri_complete")
	if verificationUriComplete == "" {
		verificationUriComplete = verificationUri
	}

	return &DeviceAuthResponse{
		DeviceCode:              getStr(data, "device_code"),
		UserCode:                getStr(data, "user_code"),
		VerificationUri:         verificationUri,
		VerificationUriComplete: verificationUriComplete,
		ExpiresIn:               expiresIn,
		Interval:                interval,
	}, nil
}

// PollDeviceToken polls the token endpoint until authorization completes or times out.
func PollDeviceToken(ctx context.Context, httpClient *http.Client, appId, appSecret string, brand core.LarkBrand, deviceCode string, interval, expiresIn int, errOut io.Writer) *DeviceFlowResult {
	if errOut == nil {
		errOut = io.Discard
	}

	const maxPollInterval = 60
	const maxPollAttempts = 200

	endpoints := ResolveOAuthEndpoints(brand)
	deadline := time.Now().Add(time.Duration(expiresIn) * time.Second)
	currentInterval := interval
	attempts := 0

	for time.Now().Before(deadline) && attempts < maxPollAttempts {
		attempts++
		if ctx.Err() != nil {
			return &DeviceFlowResult{OK: false, Error: "expired_token", Message: "Polling was cancelled"}
		}

		select {
		case <-time.After(time.Duration(currentInterval) * time.Second):
		case <-ctx.Done():
			return &DeviceFlowResult{OK: false, Error: "expired_token", Message: "Polling was cancelled"}
		}

		form := url.Values{}
		form.Set("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
		form.Set("device_code", deviceCode)
		form.Set("client_id", appId)
		form.Set("client_secret", appSecret)

		req, err := http.NewRequest("POST", endpoints.Token, strings.NewReader(form.Encode()))
		if err != nil {
			continue
		}
		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

		resp, err := httpClient.Do(req)
		if err != nil {
			fmt.Fprintf(errOut, "[lark-cli] [WARN] device-flow: poll network error: %v\n", err)
			currentInterval = minInt(currentInterval+1, maxPollInterval)
			continue
		}

		body, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			fmt.Fprintf(errOut, "[lark-cli] [WARN] device-flow: poll read error: %v\n", err)
			currentInterval = minInt(currentInterval+1, maxPollInterval)
			continue
		}

		var data map[string]interface{}
		if err := json.Unmarshal(body, &data); err != nil {
			fmt.Fprintf(errOut, "[lark-cli] [WARN] device-flow: poll parse error: %v\n", err)
			currentInterval = minInt(currentInterval+1, maxPollInterval)
			continue
		}

		errStr := getStr(data, "error")

		if errStr == "" && getStr(data, "access_token") != "" {
			fmt.Fprintf(errOut, "[lark-cli] device-flow: token obtained successfully\n")
			refreshToken := getStr(data, "refresh_token")
			tokenExpiresIn := getInt(data, "expires_in", 7200)
			refreshExpiresIn := getInt(data, "refresh_token_expires_in", 604800)
			if refreshToken == "" {
				fmt.Fprintf(errOut, "[lark-cli] [WARN] device-flow: no refresh_token in response\n")
				refreshExpiresIn = tokenExpiresIn
			}
			return &DeviceFlowResult{
				OK: true,
				Token: &DeviceFlowTokenData{
					AccessToken:      getStr(data, "access_token"),
					RefreshToken:     refreshToken,
					ExpiresIn:        tokenExpiresIn,
					RefreshExpiresIn: refreshExpiresIn,
					Scope:            getStr(data, "scope"),
				},
			}
		}

		switch errStr {
		case "authorization_pending":
			continue
		case "slow_down":
			currentInterval = minInt(currentInterval+5, maxPollInterval)
			fmt.Fprintf(errOut, "[lark-cli] device-flow: slow_down, interval increased to %ds\n", currentInterval)
			continue
		case "access_denied":
			msg := getStr(data, "error_description")
			if msg == "" {
				msg = "Authorization denied by user"
			}
			return &DeviceFlowResult{OK: false, Error: "access_denied", Message: msg}
		case "expired_token", "invalid_grant":
			msg := getStr(data, "error_description")
			if msg == "" {
				msg = "Device code expired, please try again"
			}
			return &DeviceFlowResult{OK: false, Error: "expired_token", Message: msg}
		}

		desc := getStr(data, "error_description")
		if desc == "" {
			desc = errStr
		}
		if desc == "" {
			desc = "Unknown error"
		}
		fmt.Fprintf(errOut, "[lark-cli] [WARN] device-flow: unexpected error: error=%s, desc=%s\n", errStr, desc)
		return &DeviceFlowResult{OK: false, Error: "expired_token", Message: desc}
	}

	if attempts >= maxPollAttempts {
		fmt.Fprintf(errOut, "[lark-cli] [WARN] device-flow: max poll attempts (%d) reached\n", maxPollAttempts)
	}
	return &DeviceFlowResult{OK: false, Error: "expired_token", Message: "Authorization timed out, please try again"}
}

// helpers

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func getStr(m map[string]interface{}, key string) string {
	if v, ok := m[key]; ok {
		if s, ok := v.(string); ok {
			return s
		}
	}
	return ""
}

func getInt(m map[string]interface{}, key string, fallback int) int {
	if v, ok := m[key]; ok {
		switch n := v.(type) {
		case float64:
			return int(n)
		case int:
			return n
		}
	}
	return fallback
}

File: 研究/feishu/openclaw-lark-main/src/core/device-flow.ts
/**
 * Copyright (c) 2026 ByteDance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 *
 * OAuth 2.0 Device Authorization Grant (RFC 8628) for Lark/Feishu.
 *
 * Two-step flow:
 *   1. `requestDeviceAuthorization` – obtains device_code + user_code.
 *   2. `pollDeviceToken` – polls the token endpoint until the user authorises,
 *      rejects, or the code expires.
 *
 * All HTTP calls use the built-in `fetch` (Node 18+). The Lark SDK is not
 * used here because these OAuth endpoints are outside the SDK's scope.
 */

import type { LarkBrand } from './types';
import { larkLogger } from './lark-logger';

const log = larkLogger('core/device-flow');
import { feishuFetch } from './feishu-fetch';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface DeviceAuthResponse {
  deviceCode: string;
  userCode: string;
  verificationUri: string;
  verificationUriComplete: string;
  expiresIn: number; // seconds
  interval: number; // recommended polling interval (seconds)
}

export interface DeviceFlowTokenData {
  accessToken: string;
  refreshToken: string;
  expiresIn: number; // seconds
  refreshExpiresIn: number; // seconds
  scope: string;
}

export type DeviceFlowResult =
  | { ok: true; token: DeviceFlowTokenData }
  | { ok: false; error: DeviceFlowError; message: string };

export type DeviceFlowError = 'authorization_pending' | 'slow_down' | 'access_denied' | 'expired_token';

// ---------------------------------------------------------------------------
// Endpoint resolution
// ---------------------------------------------------------------------------

/**
 * Resolve the two OAuth endpoint URLs based on the configured brand.
 */
export function resolveOAuthEndpoints(brand: LarkBrand): {
  deviceAuthorization: string;
  token: string;
} {
  if (!brand || brand === 'feishu') {
    return {
      deviceAuthorization: 'https://accounts.feishu.cn/oauth/v1/device_authorization',
      token: 'https://open.feishu.cn/open-apis/authen/v2/oauth/token',
    };
  }
  if (brand === 'lark') {
    return {
      deviceAuthorization: 'https://accounts.larksuite.com/oauth/v1/device_authorization',
      token: 'https://open.larksuite.com/open-apis/authen/v2/oauth/token',
    };
  }
  // Custom domain – derive paths by convention.
  // Smart derivation: open.X → accounts.X for the device authorization endpoint.
  const base = brand.replace(/\/+$/, '');
  let accountsBase = base;
  try {
    const parsed = new URL(base);
    if (parsed.hostname.startsWith('open.')) {
      accountsBase = `${parsed.protocol}//${parsed.hostname.replace(/^open\./, 'accounts.')}`;
    }
  } catch {
    /* fallback to base */
  }

  return {
    deviceAuthorization: `${accountsBase}/oauth/v1/device_authorization`,
    token: `${base}/open-apis/authen/v2/oauth/token`,
  };
}

// ---------------------------------------------------------------------------
// Step 1 – Device Authorization Request
// ---------------------------------------------------------------------------

/**
 * Request a device authorisation code from the Feishu OAuth server.
 *
 * Uses Confidential Client authentication (HTTP Basic with appId:appSecret).
 * The `offline_access` scope is automatically appended so that the token
 * response includes a refresh_token.
 */
export async function requestDeviceAuthorization(params: {
  appId: string;
  appSecret: string;
  brand: LarkBrand;
  scope?: string;
}): Promise<DeviceAuthResponse> {
  const { appId, appSecret, brand } = params;
  const endpoints = resolveOAuthEndpoints(brand);

  // Ensure offline_access is always requested.
  let scope = params.scope ?? '';
  if (!scope.includes('offline_access')) {
    scope = scope ? `${scope} offline_access` : 'offline_access';
  }

  const basicAuth = Buffer.from(`${appId}:${appSecret}`).toString('base64');

  const body = new URLSearchParams();
  body.set('client_id', appId);
  body.set('scope', scope);

  log.info(
    `requesting device authorization (scope="${scope}") url=${endpoints.deviceAuthorization} token_url=${endpoints.token}`,
  );

  const resp = await feishuFetch(endpoints.deviceAuthorization, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      Authorization: `Basic ${basicAuth}`,
    },
    body: body.toString(),
  });

  const text = await resp.text();
  log.info(`response status=${resp.status} body=${text.slice(0, 500)}`);

  let data: Record<string, unknown>;
  try {
    data = JSON.parse(text) as Record<string, unknown>;
  } catch {
    throw new Error(`Device authorization failed: HTTP ${resp.status} – ${text.slice(0, 200)}`);
  }

  if (!resp.ok || data.error) {
    const msg = (data.error_description as string) ?? (data.error as string) ?? 'Unknown error';
    throw new Error(`Device authorization failed: ${msg}`);
  }

  const expiresIn = (data.expires_in as number) ?? 240;
  const interval = (data.interval as number) ?? 5;
  log.info(`device_code obtained, expires_in=${expiresIn}s (${Math.round(expiresIn / 60)}min), interval=${interval}s`);

  return {
    deviceCode: data.device_code as string,
    userCode: data.user_code as string,
    verificationUri: data.verification_uri as string,
    verificationUriComplete: (data.verification_uri_complete as string) ?? (data.verification_uri as string),
    expiresIn,
    interval,
  };
}

// ---------------------------------------------------------------------------
// Step 2 – Poll Token Endpoint
// ---------------------------------------------------------------------------

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    const timer = setTimeout(resolve, ms);
    signal?.addEventListener(
      'abort',
      () => {
        clearTimeout(timer);
        reject(new DOMException('Aborted', 'AbortError'));
      },
      { once: true },
    );
  });
}

/**
 * Poll the token endpoint until the user authorises, rejects, or the code
 * expires.
 *
 * Handles `authorization_pending` (keep polling), `slow_down` (back off by
 * +5 s), `access_denied` and `expired_token` (terminal errors).
 *
 * Pass an `AbortSignal` to cancel polling from the outside.
 */
export async function pollDeviceToken(params: {
  appId: string;
  appSecret: string;
  brand: LarkBrand;
  deviceCode: string;
  interval: number;
  expiresIn: number;
  signal?: AbortSignal;
}): Promise<DeviceFlowResult> {
  const MAX_POLL_INTERVAL = 60; // slow_down 最大间隔 60 秒
  const MAX_POLL_ATTEMPTS = 200; // 安全上限（远超设备码有效期）

  const { appId, appSecret, brand, deviceCode, expiresIn, signal } = params;
  let interval = params.interval;
  const endpoints = resolveOAuthEndpoints(brand);
  const deadline = Date.now() + expiresIn * 1000;
  let attempts = 0;

  while (Date.now() < deadline && attempts < MAX_POLL_ATTEMPTS) {
    attempts++;
    if (signal?.aborted) {
      return { ok: false, error: 'expired_token', message: 'Polling was cancelled' };
    }

    await sleep(interval * 1000, signal);

    let data: Record<string, unknown>;
    try {
      const resp = await feishuFetch(endpoints.token, {
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        body: new URLSearchParams({
          grant_type: 'urn:ietf:params:oauth:grant-type:device_code',
          device_code: deviceCode,
          client_id: appId,
          client_secret: appSecret,
        }).toString(),
      });
      data = (await resp.json()) as Record<string, unknown>;
    } catch (err) {
      log.warn(`poll network error: ${err}`);
      interval = Math.min(interval + 1, MAX_POLL_INTERVAL);
      continue;
    }

    const error = data.error as string | undefined;

    if (!error && data.access_token) {
      log.info('token obtained successfully');
      const refreshToken = (data.refresh_token as string) ?? '';
      const expiresIn = (data.expires_in as number) ?? 7200;
      let refreshExpiresIn = (data.refresh_token_expires_in as number) ?? 604800;
      if (!refreshToken) {
        log.warn('no refresh_token in response, token will not be refreshable');
        refreshExpiresIn = expiresIn;
      }
      return {
        ok: true,
        token: {
          accessToken: data.access_token as string,
          refreshToken,
          expiresIn,
          refreshExpiresIn,
          scope: (data.scope as string) ?? '',
        },
      };
    }

    if (error === 'authorization_pending') {
      log.debug('authorization_pending, retrying...');
      continue;
    }

    if (error === 'slow_down') {
      interval = Math.min(interval + 5, MAX_POLL_INTERVAL);
      log.info(`slow_down, interval increased to ${interval}s`);
      continue;
    }

    if (error === 'access_denied') {
      log.info('user denied authorization');
      return { ok: false, error: 'access_denied', message: '用户拒绝了授权' };
    }

    if (error === 'expired_token' || error === 'invalid_grant') {
      log.info(`device code expired/invalid (error=${error})`);
      return { ok: false, error: 'expired_token', message: '授权码已过期，请重新发起' };
    }

    // Unknown error – treat as terminal.
    const desc = (data.error_description as string) ?? error ?? 'Unknown error';
    log.warn(`unexpected error: error=${error}, desc=${desc}`);
    return { ok: false, error: 'expired_token', message: desc };
  }

  if (attempts >= MAX_POLL_ATTEMPTS) {
    log.warn(`max poll attempts (${MAX_POLL_ATTEMPTS}) reached`);
  }
  return { ok: false, error: 'expired_token', message: '授权超时，请重新发起' };
}
